orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/gce/GoogleImageTagger.java

```diff
@@ -0,0 +1,192 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+
+@Component
+public class GoogleImageTagger implements ImageTagger, CloudProviderAware {
+  private static final Logger log = LoggerFactory.getLogger(GoogleImageTagger.class);
+  private static final Set<String> BUILT_IN_TAGS = new HashSet<>(Arrays.asList("appversion", "build_host"));
+
+  @Autowired
+  OortService oortService;
+
+  @Autowired
+  ObjectMapper objectMapper;
+
+  @Override
+  public ImageTagger.OperationContext getOperationContext(Stage stage) {
+    StageData stageData = (StageData) stage.mapTo(StageData.class);
+
+    Collection<MatchedImage> matchedImages = findImages(stageData.imageNames, stage);
+
+    stageData.imageNames = matchedImages.stream()
+      .map(matchedImage -> matchedImage.imageName)
+      .collect(Collectors.toSet());
+
+    // Built-in tags are not updatable
+    Map<String, String> tags = stageData.tags.entrySet().stream()
+      .filter(entry -> !BUILT_IN_TAGS.contains(entry.getKey()))
+      .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    List<Image> targetImages = new ArrayList<>();
+    Map<String, Object> originalTags = new HashMap<>();
+    List<Map<String, Map>> operations = new ArrayList<>();
+
+    for (MatchedImage matchedImage : matchedImages) {
+      Image targetImage = new Image(
+        matchedImage.imageName,
+        getCredentials(stage),
+        null,
+        tags
+      );
+      targetImages.add(targetImage);
+
+      log.info(format("Tagging '%s' with '%s' (executionId: %s)", targetImage.imageName, targetImage.tags, stage.getExecution().getId()));
+
+      // Update the tags on the image
+      operations.add(
+        ImmutableMap.<String, Map>builder()
+          .put(OPERATION, ImmutableMap.builder()
+            .put("imageName", targetImage.imageName)
+            .put("tags", targetImage.tags)
+            .put("credentials", targetImage.account)
+            .build()
+          ).build()
+      );
+
+      originalTags.put(matchedImage.imageName, matchedImage.tags);
+    }
+
+    Map<String, Object> extraOutput = objectMapper.convertValue(stageData, Map.class);
+    extraOutput.put("targets", targetImages);
+    extraOutput.put("originalTags", originalTags);
+
+    return new ImageTagger.OperationContext(operations, extraOutput);
+  }
+
+  /**
+   * Return true iff the tags on the current machine image match the desired.
+   */
+  @Override
+  public boolean areImagesTagged(Collection<Image> targetImages, Stage stage) {
+    Collection<MatchedImage> matchedImages = findImages(
+      targetImages.stream().map(targetImage -> targetImage.imageName).collect(Collectors.toSet()),
+      stage
+    );
+
+    AtomicBoolean isUpserted = new AtomicBoolean(true);
```

Comment:
`Atomic` is a bit of a hack but likely because it's doing a new school `forEach` that doesn't want to let you re-assign a variable.